File: 257-binary-tree-paths/binary-tree-paths.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[str]
        """
        ans = []

        def preOrder(root, path):
            if root == None:
                return

            if not root.left and not root.right:
                newArr = path[:] + [str(root.val)]
                ans.append("->".join(newArr))
            
            path.append(str(root.val))

            preOrder(root.left, path)
            preOrder(root.right, path)

            path.pop()
        
        preOrder(root, [])
        return ans
            
            

File: 257-binary-tree-paths/binary-tree-paths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    ArrayList<String> a = new ArrayList<>();
    public void build(TreeNode root, StringBuilder temp) {
        if (root == null) {
            return;
        }
        if (root.left == null && root.right == null) {
            this.a.add(new StringBuilder(temp).append(root.val).toString());
        }
        build(root.left, new StringBuilder(temp).append(root.val + "->"));
        build(root.right, new StringBuilder(temp).append(root.val + "->"));

    }
    public List<String> binaryTreePaths(TreeNode root) {
        build(root, new StringBuilder());
        return this.a;
    }
}